### PR TITLE
Declare all enums as enum class

### LINKED ISSUE #319: MSVC Code Analysis warnings suggesting `enum class`
Running MSVC's Code Analysis (Analyze -> Run Code Analysis -> On Solution) gives many warnings, most of which suggest changing `enum` to `enum class`.

Warnings filtered for `enum class` suggestions:
```
C:\Users\User\source\repos\vcpkg\installed\x64-windows\include\SDL2\SDL_rect.h(102): warning C26812: The enum type 'SDL_bool' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\nas2d-core\include\NAS2D\Signal.h(65): warning C26812: The enum type 'DiggerDirection::DiggerSelection' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\Things\Structures\Structure.h(68): warning C26812: The enum type 'Structure::StructureState' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\Things\Structures\Structure.h(72): warning C26812: The enum type 'DisabledReason' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\Things\Structures\Structure.h(79): warning C26812: The enum type 'IdleReason' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\Things\Structures\Structure.h(103): warning C26812: The enum type 'Structure::StructureClass' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\Things\Structures\Structure.h(104): warning C26812: The enum type 'ConnectorDir' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\Common.cpp(82): warning C26812: The enum type 'MineProductionRate' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\Common.cpp(232): warning C26812: The enum type 'ProductType' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\Mine.h(37): warning C26812: The enum type 'MineProductionRate' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\GraphWalker.cpp(13): warning C26812: The enum type 'Direction' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\Things\Structures\Factory.h(44): warning C26812: The enum type 'ProductType' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\Things\Structures\SeedSmelter.h(53): warning C26812: The enum type 'ResourcePool::ResourceType' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\Things\Robots\Robodigger.h(13): warning C26812: The enum type 'Direction' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\Map\TileMap.cpp(41): warning C26812: The enum type 'constants::PlanetHostility' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\Map\TileMap.cpp(395): warning C26812: The enum type 'TileMap::MouseMapRegion' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\Mine.cpp(14): warning C26812: The enum type 'Mine::OreType' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\PopulationPool.cpp(42): warning C26812: The enum type 'Population::PersonRole' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\Population\Population.cpp(83): warning C26812: The enum type 'Population::PersonRole' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\ProductPool.cpp(16): warning C26812: The enum type 'ProductType' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\ResourcePool.cpp(111): warning C26812: The enum type 'ResourcePool::ResourceType' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\RobotPool.cpp(63): warning C26812: The enum type 'RobotType' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\States\MapViewState.h(220): warning C26812: The enum type 'InsertMode' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\States\MapViewState.h(221): warning C26812: The enum type 'StructureID' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\States\MapViewState.h(222): warning C26812: The enum type 'RobotType' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\States\MapViewState.cpp(95): warning C26812: The enum type 'constants::PlanetHostility' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\States\MapViewState.cpp(134): warning C26812: The enum type 'MapViewState::PopulationLevel' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\States\Planet.h(36): warning C26812: The enum type 'Planet::PlanetType' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\States\MainMenuState.cpp(169): warning C26812: The enum type 'FileIo::FileOperation' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\States\MapViewStateHelper.cpp(320): warning C26812: The enum type 'StructureID' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\States\MapViewStateHelper.cpp(534): warning C26812: The enum type 'RobotType' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\States\MapViewStateIO.cpp(362): warning C26812: The enum type 'StructureID' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\States\MapViewStateUi.cpp(446): warning C26812: The enum type 'RobotType' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\States\MapViewStateUi.cpp(448): warning C26812: The enum type 'InsertMode' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\States\MapViewStateUi.cpp(456): warning C26812: The enum type 'DiggerDirection::DiggerSelection' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\States\MapViewStateUi.cpp(561): warning C26812: The enum type 'FileIo::FileOperation' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\States\PlanetSelectState.cpp(190): warning C26812: The enum type 'constants::PlanetHostility' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\states\SplashState.cpp(34): warning C26812: The enum type 'LogoState' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\StructureCatalogue.cpp(25): warning C26812: The enum type 'StructureID' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\StructureTranslator.cpp(17): warning C26812: The enum type 'StructureID' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\UI\Core\Button.h(58): warning C26812: The enum type 'Button::State' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\UI\Core\Button.h(59): warning C26812: The enum type 'Button::Type' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\UI\Core\Slider.cpp(58): warning C26812: The enum type 'Slider::SliderType' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\UI\Core\TextField.h(79): warning C26812: The enum type 'TextField::BorderVisibility' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\ui\FileIo.cpp(109): warning C26812: The enum type 'FileIo::FileOperation' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\UI\MajorEventAnnouncement.cpp(44): warning C26812: The enum type 'MajorEventAnnouncement::AnnouncementType' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
Z:\OPHD\src\UI\ResourceBreakdownPanel.cpp(25): warning C26812: The enum type 'ResourceTrend' is unscoped. Prefer 'enum class' over 'enum' (Enum.3).
```

----

As a side note, some of those paths have inconsistencies in the casing. In particular:
- "states" in "Z:\OPHD\src\states\SplashState.cpp"
- "io" in "Z:\OPHD\src\ui\FileIo.cpp"

### PR DESCRIPTION
**Pull request by cugone**
...Then fix the *thousands* of errors that result.

Closes #319

### CODE REVIEW

**Comment by cugone** [comment]
Weird. I didn't change any whitespace intentionally. Might have been a tidy collision.

I'll go back through and fix up any formatting/whitespace issues as a separate PR. I don't like mixing whitespace fixes with actual changes.

File: OPHD/Mine.cpp [hunk 1]
```diff
@@ -16,18 +16,18 @@ static int getOreCount(const Mine::MineVeins& veins, Mine::OreType ore, size_t d
...
 static void setDefaultFlags(std::bitset<6>& flags)
 {
-	flags[Mine::ORE_COMMON_METALS] = true;
-	flags[Mine::ORE_COMMON_MINERALS] = true;
-	flags[Mine::ORE_RARE_METALS] = true;
-	flags[Mine::ORE_RARE_MINERALS] = true;
+	flags[static_cast<std::size_t>(Mine::OreType::ORE_COMMON_METALS)] = true;
+	flags[static_cast<std::size_t>(Mine::OreType::ORE_COMMON_MINERALS)] = true;
+	flags[static_cast<std::size_t>(Mine::OreType::ORE_RARE_METALS)] = true;
+	flags[static_cast<std::size_t>(Mine::OreType::ORE_RARE_MINERALS)] = true;
```
Comment (DanRStevens) [inline OPHD/Mine.cpp:30]: The large number of `static_cast` that were introduced should perhaps have been brought up as a refactoring problem.

It would be nice to find a way to update that doesn't involve `static_cast`. That might mean a code structure change, or it might mean providing an operator on the `enum class` so that it auto converts to `std::size_t`. It should be possible to allow that to happen implicitly.

Comment (cugone) [reply OPHD/Mine.cpp:30]: Will also need to add conversion operators for `int` (`operator int`) as well as `std::size_t` (`operator std::size_t`). In several places the enums were used as arguments to index functions that took `int`s.

Comment (DanRStevens) [reply OPHD/Mine.cpp:30]: We might also investigate if those functions should take `std::size_t` instead.

File: OPHD/States/MapViewStateHelper.cpp [hunk 2]
```diff
@@ -402,7 +402,7 @@ void transferProductsPool(ProductPool& source, ProductPool& destination)
 
 	auto& src = source.mProducts;
 
-	for (size_t i = 0; i < PRODUCT_COUNT; ++i)
+	for (std::size_t i = 0; i < static_cast<std::size_t>(ProductType::PRODUCT_COUNT); ++i)
```
Comment (DanRStevens) [inline OPHD/States/MapViewStateHelper.cpp:405]: The `std::` prefix on `std::size_t` is good, but should probably be applied in a separate commit.

File: OPHD/Templates.h [hunk 3]
```diff
@@ -19,7 +19,7 @@ void transferProductsStructure(T& source, T& destination)
 	auto& src = source->products().mProducts;
 	auto& dest = destination->products();
 
-	for (size_t i = 0; i < PRODUCT_COUNT; ++i)
+	for (std::size_t i = 0; i < static_cast<std::size_t>(ProductType::PRODUCT_COUNT); ++i)
```
Comment (DanRStevens) [inline OPHD/Templates.h:22]: Ditto.

File: OPHD/UI/FactoryProduction.cpp [hunk 4]
```diff
@@ -196,12 +196,12 @@ void FactoryProduction::factory(Factory* newFactory)
 		throw std::runtime_error("FactoryProduction::factory(): Factory provided with an empty production type list.");
 	}
 
-	for (size_t i = 0; i < ptlist.size(); ++i)
+	for (std::size_t i = 0; i < ptlist.size(); ++i)
```
Comment (DanRStevens) [inline OPHD/UI/FactoryProduction.cpp:199]: Ditto.

File: OPHD/UI/ProductListBox.cpp [hunk 5]
```diff
@@ -73,7 +73,7 @@ void ProductListBox::productPool(ProductPool& pool)
 {
 	clearItems();
 
-	for (size_t product_type = 0; product_type < static_cast<size_t>(PRODUCT_COUNT); ++product_type)
+	for (std::size_t product_type = 0; product_type < static_cast<size_t>(ProductType::PRODUCT_COUNT); ++product_type)
```
Comment (DanRStevens) [inline OPHD/UI/ProductListBox.cpp:76]: Ditto.

**Comment by cugone** [comment]
@cugone Tagging myself to remind me with an email notify to look at this later, it's 2 am. -.-
